### PR TITLE
Add referrer and validate signature

### PR DESCRIPTION
**Pull request by bigdefidev**
this has all requested changes - not sure if str(signature) returns a hex representation but it does seem to work correctly. Thanks!

Specific changes can be found here: https://friktionlabsgroup.slack.com/archives/C03DTHWFJAX/p1658860434461269?thread_ts=1658858875.586349&cid=C03DTHWFJAX

Summary:
1. validate signature in validate_bid\
2. add referrer to bid_details

### CODE REVIEW

**Comment by pdonorio** [comment]
@bigdefidev thanks for this!

Can we give a better title and list in the description the things being solved?
It would make reviewing faster.

File: friktion/friktion/bid_details.py [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
         byte_rep = b"".join(
             [
                 bytes([self.order_id]),
                 wallet.public_key.__bytes__(),
+                self.referrer.__bytes__(),
                 bytes([give_amount, receive_amount]),
             ]
         )
-        return wallet.payer.sign(byte_rep)
+        return (byte_rep, wallet.payer.sign(byte_rep))
```
Comment (cassios) [inline friktion/friktion/bid_details.py:28]: the return of `wallet.payer.sign(byte_rep)` is not of type `Signature`


Screen Shot 2022-07-28 at 10 04 00

File: friktion/main.py [hunk 2]
```diff
@@ -111,13 +115,17 @@ async def main_def():
...
         signer_wallet=COUNTERPARTY,
+        referrer=REFERRER
     )
+
+    # happens outside of paradigm
+    (msg, signature) = bid_details.as_signed_msg(wallet)
+
     # fill offer via bid
-    if error := await c.validate_bid(swap_order_creator, bid_details):
+    if error := await c.validate_bid(swap_order_creator, bid_details, msg, str(signature)):
```
Comment (cassios) [inline friktion/main.py:125]: `str(signature)` is wrapping the bytes for the message and the bytes for signature itself into a string. We won't be able to do the same in our workflow as we only expect a hexadecimal string from our users on `/quote` endpoint.

Screen Shot 2022-07-28 at 10 29 09

Screen Shot 2022-07-27 at 10 07 58

Comment (cassios) [reply friktion/main.py:125]: getting this error when trying to run the script
Screen Shot 2022-07-28 at 10 44 48

**Review by cassios — COMMENTED** [review]
Thanks for opening the PR. Added some comments

**Comment by bigdefidev** [comment]
> @bigdefidev thanks for this!
> 
> Can we give a better title and list in the description the things being solved? It would make reviewing faster.
> 
> Edit: also linters are failing, see https://github.com/tradeparadigm/sdks#code-guidelines

updated! Wasn't sure how to run the linter but checking it out now

File: friktion/friktion/bid_details.py (hunk 1 above)
Comment (bigdefidev) [reply friktion/friktion/bid_details.py:28]: Weird, that's what I'm seeing locally, and mypy isn't catching the error

Comment (bigdefidev) [reply friktion/friktion/bid_details.py:28]: image

might have been fixed in my newest updates?

File: friktion/friktion/swap.py [hunk 3]
```diff
@@ -201,8 +203,12 @@ async def _validate_bid_allowance(
         if acct_info.amount < transfer_amount:
             return "amount in token account is below required threshold"
 
+        return None
+
     async def validate_bid(
-        self, swap_order_creator: PublicKey, bid_details: BidDetails
+        self, swap_order_creator: PublicKey, bid_details: BidDetails,
+        message: bytes,
+        signature: str
```
[the remark this answers is not in the document]
Comment (bigdefidev) [reply friktion/friktion/swap.py]: roger, fixed!

File: friktion/friktion/bid_details.py [hunk 4]
```diff
@@ -1,25 +1,28 @@
 from dataclasses import dataclass
+from typing import Tuple
 
 from anchorpy import Wallet
 from nacl import signing  # type: ignore
 from solana.publickey import PublicKey
-
+from solders.signature import Signature
```
[the remark this answers is not in the document]
Comment (bigdefidev) [reply friktion/friktion/bid_details.py:6]: added!

File: friktion/main.py (hunk 2 above)
Comment (bigdefidev) [reply friktion/main.py]: can you try one more time for me? I can't recreate, might have been my dependency management when testing in initial PR

**Comment by bigdefidev** [comment]
> @bigdefidev thanks for this!
> 
> Can we give a better title and list in the description the things being solved? It would make reviewing faster.
> 
> Edit: also linters are failing, see https://github.com/tradeparadigm/sdks#code-guidelines

fixed the lint errors!

File: friktion/friktion/bid_details.py (hunk 1 above)
Comment (cassios) [reply friktion/friktion/bid_details.py]: Maybe you are missing to push some changes?
I checkout out to the newest commit `3e978027bcf8` but I am still getting `SignedMessage` as the return of `as_signed_msg`

Screen Shot 2022-07-29 at 08 53 36
Screen Shot 2022-07-29 at 08 56 42

Comment (cassios) [reply friktion/friktion/bid_details.py]: My bad. Updating solana and anchorpy fixes the issue

**Review by cassios — COMMENTED** [review]
Nice. It is looking good 
I am talking to our frontend engineers about the signature using Phantom and I will get back here soon.

**Comment by cassios** [comment]
I will merge this PR, so I can integrate the changes. If any adjustment is needed, we can create a small PR for that

**Review by cassios — APPROVED** [review]

**Pull request merged**